executor/src/witgen/affine_expression.rs

```diff
@@ -36,34 +38,108 @@ where
 {
     pub fn from_variable_id(var_id: K) -> AffineExpression<K, T> {
         Self {
-            coefficients: BTreeMap::from([(var_id, T::one())]),
+            coefficients: vec![(var_id, T::one())],
             offset: T::zero(),
+            clean: true,
         }
     }
 
     pub fn is_constant(&self) -> bool {
-        self.nonzero_coefficients().next().is_none()
+        self.nonzero_coefficients().is_empty()
     }
 
     pub fn constant_value(&self) -> Option<T> {
         self.is_constant().then_some(self.offset)
     }
 
     pub fn nonzero_variables(&self) -> Vec<K> {
-        self.nonzero_coefficients().map(|(i, _)| i).collect()
+        self.nonzero_coefficients()
+            .into_iter()
+            .map(|(i, _)| i)
+            .collect()
     }
 
-    /// @returns an iterator of the nonzero coefficients and their variable IDs (but not the offset).
-    pub fn nonzero_coefficients(&self) -> impl Iterator<Item = (K, &T)> {
-        self.coefficients
-            .iter()
-            .filter_map(|(i, c)| (!c.is_zero()).then_some((*i, c)))
+    /// @returns an vector of the nonzero coefficients and their variable IDs (but not the offset).
+    pub fn nonzero_coefficients(&self) -> Vec<(K, T)> {
+        // We need to make sure that there are no duplicates in the variable
+        // IDs and that the coefficients are nonzero. In other words, we need to
+        // "clean" the coefficient array.
+
+        // First try the easy cases.
+        if self.clean {
+            return self.coefficients.clone();
+        }
+
+        match &self.coefficients[..] {
+            [] => return vec![],
+            [(k, v)] => return Self::clean_one(k, v),
+            [(k1, v1), (k2, v2)] => return Self::clean_two((k1, v1), (k2, v2)),
+            _ => {}
+        };
+
+        // Ok, this is more complicated.
+        // Remove duplicates by first sorting and then going through
+        // all adjacent pairs with equal variable IDs, adding the coefficient
+        // of the first to the second, and setting the coefficient of the
+        // first to zero.
+        // Then we filter out the zeros as a last step.
+        let mut coefficients = self.coefficients.clone();
+        coefficients.sort_unstable_by(|(k1, _), (k2, _)| k1.cmp(k2));
+        for i in 1..coefficients.len() {
+            let (first, second) = coefficients.split_at_mut(i);
+            let (k1, v1) = first.last_mut().unwrap();
+            let (k2, v2) = second.first_mut().unwrap();
+            if k1 == k2 {
+                *v2 += *v1;
+                *v1 = 0.into();
+            }
+        }
+        coefficients
+            .into_iter()
+            .filter(|(_, c)| !c.is_zero())
+            .collect()
+    }
```

Comment:
Is folding into a map and back out too expensive? Something like

```
coefficients.into_iter().fold(BTreeMap::default(), |mut map, (k, v)| {
   map.entry(k).or_default() += v;
   map
}).into_iter().collect()
```

Reply:
Yes, this is another 5-10 percent.

Reply:
This is the best solution I found for iterating with a mutable window of size 2...